### PR TITLE
qa: add qa/tasks/nvmeof.py

### PR DESCRIPTION
**Pull request by VallariAg**
This PR introduces version 2 of nvmeof teuthology test.

This is v2 of the rbd/nvmeof test: It deploys 1 gateway and 1 initiator.
Then does basic verification on nvme commands and runs fio.

This PR creates:
1. qa/tasks/nvmeof.py: adds a new 'Nvmeof' task which deploys
    the gateway and shares config with the initiator hosts.
    Sharing config was previously done by 'nvmeof_gateway_cfg' task
    in qa/tasks/cephadm.py (that task is removed in this commit).
2. qa/workunits/rbd/nvmeof_basic_tests.sh:
    Runs nvme commands (discovery, connect, connect-all, disconnect-all,
    and list-subsys) and does basic verification of the output.
3. qa/workunits/rbd/nvmeof_fio_test.sh:
    Runs fio command. Also runs iostat in parallel if IOSTAT_INTERVAL
    variable is set. This variable configures the delay between each iostat
    print.

nvmeof-cli upgrade from v0.0.6 to v0.0.7 introduced major changes
to all nvmeof commands. This commit changes v0.0.6 commands to
v0.0.7 in qa/workunits/rbd/nvmeof_initiator.sh

Output run: https://pulpito.ceph.com/vallariag-2024-02-09_15:14:45-rbd:nvmeof-main-distro-default-smithi/

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by zmc — CHANGES_REQUESTED** [review]
Looks very good! I had two very small suggestions.

**Comment by VallariAg** [comment]
@zmc thank you for reviewing this! I have put the fio version on top as a constant all-caps variable and added an option to specify fio version in job yaml config.

File: qa/tasks/nvmeof.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+                ioengine_pkg = 'libaio-devel' if system_type == 'rpm' else 'libaio-dev'
+                log.info("[nvmeof]: Installing {ioengine_pkg} on {system_type}".format(ioengine_pkg=ioengine_pkg, system_type=system_type))
+                install_package(ioengine_pkg, self.remote)
+            fio_config = self.write_fio_config(config)
+
+            fio = "https://github.com/axboe/fio/archive/fio-" + fio_version + ".tar.gz"
+            self.remote.run(args=['mkdir', run.Raw(test_dir),])
+            self.remote.run(args=['cd' , run.Raw(test_dir),
+                             run.Raw(';'), 'wget', fio, run.Raw(';'), run.Raw('tar -xvf fio*tar.gz'), run.Raw(';'),
+                             run.Raw('cd fio-fio*'), run.Raw(';'), './configure', run.Raw(';'), 'make'])
```
Comment (idryomov) [inline qa/tasks/nvmeof.py:374]: > I have put the fio version on top as a constant all-caps variable and added an option to specify fio version in job yaml config.

Why is `fio` being built from scratch at all?  Packages should be universally available these days -- it doesn't look like the most recent version is needed here.

It looks like this was copied from `qa/tasks/rbd_fio.py` -- I think the reason it's done there is that in 2015 the RBD engine wasn't packaged ;)  For NVMeOF, `--ioengine=rbd` won't ever be used, so it's moot point.

We should probably get rid of that and just install a package in `qa/tasks/rbd_fio.py` too.

Comment (VallariAg) [reply qa/tasks/nvmeof.py:374]: I saw `fio` being built from scratch in `qa/tasks/cbt.py` as well (this is also an old commit so it must be for the same reason): 
https://github.com/ceph/ceph/blob/19aa8fc73d267a722c0899bf830bff671b921aa7/qa/tasks/cbt.py#L108

I'll install the packaged fio for the nvmeof task. And maybe open separate PR to change it for `qa/tasks/rbd_fio.py` and `qa/tasks/cbt.py`. Thank you for pointing this out!

Comment (caroav) [reply qa/tasks/nvmeof.py]: @idryomov interesting I didn't know we have rbd_fio. Can you say few words about how it is different from fio? And BTW we're looking for running fio on many namespaces in parallel, with data verification, R/W, etc, to test HA. Do you have some examples that we look at how to do it? thx. (@barakda )

Comment (idryomov) [reply qa/tasks/nvmeof.py]: > Can you say few words about how it is different from fio?

fio has a concept of an I/O engine, which defines the I/O and how it's issued.  Think of it is as interface which gets implemented by multiple backends, which could be file-based, block device-based, network-based or even virtual.  With `--ioengine=sync` requests are issued using regular `read` and `write` syscalls, with `--ioengine=libaio` using `io_submit` and `io_getevents` syscalls and related helpers from libaio, etc.   For `--ioengine=rbd`, it's `rbd_aio_read()` and `rbd_aio_write()` APIs from librbd -- direct to the cluster with no block device or any kind of gateway.

> And BTW we're looking for running fio on many namespaces in parallel, with data verification, R/W, etc, to test HA. Do you have some examples that we look at how to do it?

I think you would need to define several fio jobs.  I'd suggest perusing https://fio.readthedocs.io/en/latest/fio_doc.html for details.

**Review by idryomov — COMMENTED** [review]
@caroav asked me to take a deeper look at this, but I think I'm missing some context and this isn't mentioned in the commit message.

What is the motivation for wrapping host (initiator)-side steps/tests which are inherently CLI-based into a teuthology task in Python?  The Python code here just executes CLI commands in a synchronous fashion and does trivial tests on their output, which would take less effort/lines of code in a shell script.  Not to mention that `qa/workunits/rbd/nvmeof_initiator.sh` script is still there and the pool and image names have to be passed to both the task and the script.  The job flow is weird: first the task sets up the gateway, then the script sets up the initiator, and then the task takes over to run tests on the initiator.

Is there a problem with simply adding new tests to `qa/workunits/rbd/nvmeof_initiator.sh`?

**Comment by caroav** [comment]
@idryomov - I think that the reason for doing things in python, is that eventually we would like to reuse the testing code, also in our local automation testing. Maybe @VallariAg or @barakda can check if they agree with me here.

**Comment by VallariAg** [comment]
Let me provide the context I'm working with, please let me know if I can do something better here:  

I understand the flow is little clunky - it seems like there's too many tasks (esp when they all need same context vars) for something that can be wrapped in one shell script. The idea is gateway deployment and nvmeof_initiator.sh should be executed once. But tasks like `nvmeof.basic_tests` and `nvmeof.fio` can be re-run. For example, we have a test planned in which we'll restart all the gateways. So we can run `nvmeof.basic_tests` after that to check if things are okay after restart. Another example is running these tasks for 2 initiator setup, where we'll check if nvme commands and fio are going well on both initiator hosts. That's why those basic tests are not part of the initiator setup shell script anymore. 
I could put initiator setup commands in `nvmeof` task (which deploys the gateway), but since they'll be run once I thought they're best left to the shell script. But to make the flow neater, initiator setup can be move them to the `nvmeof` task. It might actually be helpful when we have to pass context for 2 gateways.

Apart from what Aviv mentioned about reusing code for local automation testing, another reason why these tasks are moved to python is that - with python, we can get better exception messages for these nvme command verifications. `NvmeCommands` class is a helper class with which we can do quick nvme commands + verification. You can see these being reused in FIO test too. 
Even if these basic tests were part of a different shell script, they would still need all the context passing. In fact, almost all basic_tests only need the .env file context. `RBD_POOL` is only needed for `test_device_size`.

**Comment by idryomov** [comment]
> Let me provide the context I'm working with, please let me know if I can do something better here:
> 
> I understand the flow is little clunky - it seems like there's too many tasks (esp when they all need same context vars) for something that can be wrapped in one shell script. The idea is gateway deployment and nvmeof_initiator.sh should be executed once. But tasks like `nvmeof.basic_tests` and `nvmeof.fio` can be re-run.

The same goes for workunits/shell scripts.

> For example, we have a test planned in which we'll restart all the gateways. So we can run `nvmeof.basic_tests` after that to check if things are okay after restart.

Here, restarting gateways makes sense as subtasks (one for stopping or killing abruptly, another for restarting), because it's server-side state.  (Re-)running a test, whether it's `basic_tests` or `fio` -- not so much, at least to me.  This could be achieved with something like:

```
- workunit:
    clients:
      client.1:
        - rbd/nvmeof_basic_tests.sh

- nvmeof.stop:

- sleep:
    duration: 30

- nvmeof.start:

- workunit:
    clients:
      client.1:
        - rbd/nvmeof_basic_tests.sh
```

> Another example is running these tasks for 2 initiator setup, where we'll check if nvme commands and fio are going well on both initiator hosts. That's why those basic tests are not part of the initiator setup shell script anymore.

Workunits can be executed on multiple nodes in parallel.  For example:

```
- workunit:
    clients:
      all:
        - rbd/nvmeof_basic_tests.sh
```

or

```
- workunit:
    clients:
      client.1:
        - rbd/nvmeof_basic_tests.sh
      client.2:
        - rbd/nvmeof_fio_test.sh
```

This is built into the workunit task.  On top of that, teuthology has more advanced support for grouping and parallel execution, see https://github.com/ceph/teuthology/blob/main/teuthology/task/sequential.py and https://github.com/ceph/teuthology/blob/main/teuthology/task/parallel.py.

> I could put initiator setup commands in `nvmeof` task (which deploys the gateway), but since they'll be run once I thought they're best left to the shell script. But to make the flow neater, initiator setup can be move them to the `nvmeof` task. It might actually be helpful when we have to pass context for 2 gateways.

You could put it in a subtask, like `nvmeof.populate_initiator` and invoke it explicitly just once.

> Apart from what Aviv mentioned about reusing code for local automation testing

For local testing, I think one would be able to reuse a shell script which e.g. launches fio with specific options much easier than a piece of Python which is coded to be a task and is expected to be invoked a certain way.  At a minimum, they would need to have a different piece of Python for setting up the entry points and also adjust for differences in the execution environment.

> another reason why these tasks are moved to python is that - with python, we can get better exception messages for these nvme command verifications. `NvmeCommands` class is a helper class with which we can do quick nvme commands + verification. You can see these being reused in FIO test too.

If you are referring to 

> assert expected_connect_stdout in stdout, f"Expected stdout: {expected_connect_stdout}, but got: {stdout}"

messages, you can get the same thing in shell by enabling tracing (`set -x`) and simply comparing strings.

Anyway, I just wanted to note that there are alternatives to cramming everything into a task.  If you feel that is justified or have further plans regarding the task, I wouldn't object.  In fact, there is nothing that needs blessing from RBD here -- @ceph/rbd is tagged just because of the location of `qa/workunits/rbd/nvmeof_initiator.sh`, so you can just ignore that.

**Comment by VallariAg** [comment]
@idryomov Thank you for explaining this in amazing detail. The gateway subtasks like `nvmeof`, `nvmeof.populate_initiator`, `nvmeof.stop` and `nvmeof.start` are helpful ideas - noted!

> For local testing, I think one would be able to reuse a shell script which e.g. launches fio with specific options much easier than a piece of Python which is coded to be a task and is expected to be invoked a certain way. At a minimum, they would need to have a different piece of Python for setting up the entry points and also adjust for differences in the execution environment.

That's a really good point, I'll discuss that with the team.

It makes sense to at least move fio test to shell scripts. For basic_tests, it depends on how complex these nvmeof teuthology tests would look like in future (I think having functions like `NvmeCommands(remote).run_connect_all()` to reuse could be useful for future tasks) so I'll discuss with @caroav about future tests and the local automation setup, then update you about it.

**Comment by VallariAg** [comment]
@idryomov I've created shell scripts for both the tests - basic tests and fio. We'll be reusing these scripts in the other automation environment too. Thank you so much for suggesting this change and taking out time to explain it!

**Review by barakda — APPROVED** [review]

File: qa/workunits/rbd/nvmeof_fio_test.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+rw=${RW:-randrw}
+filename=$(echo "$drives_list" | tr '\n' ':' | sed 's/:$//')
+verify=md5
+verify_fatal=1
+EOF
+
+fio --showcmd $fio_file
+sudo fio $fio_file &
+
+if [ -n "$IOSTAT_INTERVAL" ]; then
```
Comment (barakda) [inline qa/workunits/rbd/nvmeof_fio_test.sh:30]: suggesting to stop the iostat print when the fio runtime is up, there are iostat print with no IO at the end

**Review by barakda — APPROVED** [review]

File: qa/workunits/rbd/nvmeof_basic_tests.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
+set -x
```
Comment (barakda) [inline qa/workunits/rbd/nvmeof_basic_tests.sh:1]: make it:
#!/bin/bash

set -ex

File: qa/workunits/rbd/nvmeof_fio_test.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
+set -ex
```
Comment (barakda) [inline qa/workunits/rbd/nvmeof_fio_test.sh:1]: add
#!/bin/bash

**Review by barakda — APPROVED** [review]

**Review by caroav — APPROVED** [review]

**Comment by VallariAg** [comment]
jenkins test dashboard

**Comment by VallariAg** [comment]
jenkins test api

**Pull request merged**